### ISSUE TITLE
Service descriptors defining the difference between a required parameter and an optional parameter

### ISSUE THREAD

**Issue by robyww** [ivoa-std/DataLink#51]
When a service descriptor defines a parameter there seems to be no way to know if the parameter is require or if it is optional. This is important for a UI that reads the service descriptor and provides the user a way input those parameters.

This is not a suggestion of syntax, but just an illustration of a solution. I added an `isOptional` in the third parameter below.
```xml
  <RESOURCE type="meta" utype="adhoc:service">
    <DESCRIPTION>a service</DESCRIPTION>
    <GROUP name="inputParams">
      <PARAM name="abc" datatype="char" arraysize="*" ref="col_0" value="">
        <DESCRIPTION>a ref param</DESCRIPTION>
      </PARAM>
      <PARAM name="efg" datatype="char" arraysize="*"  value="123">
        <DESCRIPTION>a required param</DESCRIPTION>
      </PARAM>
      <PARAM name="mn" datatype="char" arraysize="*"  value=""  isOptional="true">
        <DESCRIPTION>a optional param</DESCRIPTION>
      </PARAM>
    </GROUP>
    <PARAM name="accessURL" datatype="char" arraysize="*" value="http://a.b.com/aEndpoint"/>
  </RESOURCE>

```

**Comment by Bonnarel** [comment]
This is surely useful. I see a solution different from adding a new attribute to PARAM. FIELDS and PARAMS have a rather seldom used attribute called "type" (no datatype, utype or xtype). It's used by clients as a behavior guid. Current values are "hidden", "location", "trigger". We could add type="optional" and that's it.
However the status of this attribute in VOTable is rather unclear. This could be revisted and clarified in VOTable. What do you think ?

**Comment by robyww** [comment]
Yes, that sounds good. I am sure you know better how to implement it than I do. As long as there is a way to determine an optional parameter.

**Comment by cmzwolf** [comment]
In PDL, the parameter description comes with an attribute "dependency" which can takes REQUIRED or OPTIONAL value (cf. section 5 of https://ivoa.net/documents/PDL/20140523/REC-PDL-1.0-20140523.pdf). What about a better integration between PDL and DataLink?

**Comment by pdowler** [comment]
This comes  up in even the simplest use cases and I'm a little embarrassed we didn't think of it. The other potentially related thing we cannot currently describe is the cardinality: is a param single-valued or can multiple be submitted in the request?

In a SODA service descriptor, one could use a static template and list all the params: pos, circle, polygon, band, time, pol. Users could use the ones they like, so I think we implicitly assumed that all params were "optional" in this sense. That certainly fits with the fact that you woould not use both POS and CIRCLE at the same time and services could (CADC SODA does) reject that kind of repeated usage (multiple spatial cutouts). 

So, to me the question could be reduced to "which params are mandatory?" I looked at the VOTable schema for attributes we could (ab)use, but there isn't anything obvious. We do use `utype="adhoc:service"` in the resource so we could think about using `PARAM utype="something"` to indicate mandatory vs optional params (with default when absent: optional). I don't think service params obviously benefit from referring to a field in a model (the nominal utype attr usage)... 

Option 1: abuse utype attribute to say required (vs optional), eg `utype="adhoc:required"`

Option 2: abuse utype to specify cardinality: `utype="cardinality:0..1"` for optional & single value, then `1..1`, `0..*`, and `1..*` for the other possibilities. 

Option 3: defer issue  until some useful support/param refactoring is added to VOTable 

I'm inclined to defer rather than abuse current votable mechanisms.

**Comment by mbtaylor** [comment]
I've made a suggestion in https://github.com/ivoa-std/VOTable/issues/29 that could provide a clean solution to this; but it's not a quick fix, since it would require a new VOTable version.
